src/converse-httpFileUpload.js

```diff
@@ -0,0 +1,269 @@
+
+    Permission is hereby granted, free of charge, to any person obtaining a copy
+    of this software and associated documentation files (the "Software"), to deal
+    in the Software without restriction, including without limitation the rights
+    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+    copies of the Software, and to permit persons to whom the Software is
+    furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included in
+    all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+    THE SOFTWARE.
+*/
+
+/**
+ * Implements Http File Upload (XEP-0363)
+ *
+ * @see {@link http://xmpp.org/extensions/xep-0363.html}
+ */
+(function (root, factory) {
+    define([
+        "converse-core",
+    ], factory);
+}(this, function (
+        converse
+    ) {
+    "use strict";
+    const { $msg, Backbone, Strophe, _, b64_sha1, moment, utils } = converse.env;
+
+    Strophe.addNamespace('HTTPUPLOAD', 'urn:xmpp:http:upload');
+
+    var requestSlotUrl; 
+    var ready;
+    var httpUploadOption = {
+        enable: true
+    }
+
+    converse.plugins.add('converse-httpFileUpload', {
+
+        dependencies: ["converse-chatboxes", "converse-disco"],
+
+        initialize() {
+            const { _converse } = this,
+                { __ } = _converse;
+            var connection = _converse.connection;
+            var domain;
+            var file;
+            var chatBox;
+            
+            _converse.FileUpload = Backbone.NativeView.extend({
+                /**
+                * Set up http file upload.
+                * 
+                * @param {*} connection the current strophe-connection
+                */
+                initFiletransfer () {
+                    connection = _converse.connection;
+                    domain = _converse.connection.domain; 
+
+                    if (httpUploadOption && requestSlotUrl != undefined) {
+                        ready = true;
+                        return;
+                    }
+                    this.discoverUploadService();
+                },
+                
+                /**
+                * Discover upload service for http upload.
+                *
+                */
+                discoverUploadService () {
+                    var self = this;
+                    console.log('discover http upload service');
+                    connection.disco.items(domain, null, function(items) {
```

Comment:
Converse.js has it's own, higher-level and Promise-based, service discovery API, exposed in `src/converse-disco.js`, and we should rather use that than the one from Strophe.

By using Converse's API, we don't need this `discoverUploadService` method at all, since that happens automatically.

Instead, we can simply call 

```
_converse.api.disco.supports(Strophe.NS.HTTPUPLOAD, _converse.domain)
    .then((result) => {
        // Your code here
    });
```

whenever we want to know whether the service is available.

See this example in the `http-file-upload` branch I've been working on:
https://github.com/conversejs/converse.js/blob/720ae390eeb1c743308c94f0ef5e1e476f76abcf/src/converse-http-file-upload.js#L40